src/Roslyn.Diagnostics.Analyzers/Core/SymbolIsBannedAnalyzer.cs

```diff
@@ -0,0 +1,231 @@
+            id: RoslynDiagnosticIds.DuplicateBannedSymbolRuleId,
+            title: RoslynDiagnosticsAnalyzersResources.DuplicateBannedSymbolTitle,
+            messageFormat: RoslynDiagnosticsAnalyzersResources.DuplicateBannedSymbolMessage,
+            category: "ApiDesign",
+            defaultSeverity: DiagnosticHelpers.DefaultDiagnosticSeverity,
+            isEnabledByDefault: DiagnosticHelpers.EnabledByDefaultIfNotBuildingVSIX,
+            description: RoslynDiagnosticsAnalyzersResources.DuplicateBannedSymbolDescription,
+            customTags: WellKnownDiagnosticTags.Telemetry);
+
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+            ImmutableArray.Create(SymbolIsBannedRule, DuplicateBannedSymbolRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            // TODO: Make the analyzer thread-safe.
+            //context.EnableConcurrentExecution();
+
+            // Analyzer needs to get callbacks for generated code, and might report diagnostics in generated code.
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private void OnCompilationStart(CompilationStartAnalysisContext compilationContext)
+        {
+            var additionalFiles = compilationContext.Options.AdditionalFiles;
+
+            if (!TryGetApiData(additionalFiles, compilationContext.CancellationToken, out ApiData apiData))
+            {
+                return;
+            }
+
+            if (!ValidateApiFiles(apiData, out List<Diagnostic> errors))
+            {
+                compilationContext.RegisterCompilationEndAction(context =>
+                {
+                    foreach (Diagnostic cur in errors)
+                    {
+                        context.ReportDiagnostic(cur);
+                    }
+                });
+
+                return;
+            }
+
+            var bannedApis = ImmutableHashSet.CreateBuilder<ISymbol>();
+            foreach (var apiString in apiData.ApiList)
+            {
+                var symbol = compilationContext.Compilation.GetTypeByMetadataName(apiString.Text);
+                if (!(symbol is null))
+                {
+                    bannedApis.Add(symbol);
+                }
+            }
+
+            if (bannedApis.Count > 0)
+            {
+                var symbolDisplayFormat = compilationContext.Compilation.Language == LanguageNames.CSharp
+                    ? SymbolDisplayFormat.CSharpShortErrorMessageFormat
+                    : SymbolDisplayFormat.VisualBasicShortErrorMessageFormat;
+                compilationContext.RegisterOperationAction(
+                    oac => AnalyzeOperation(oac, bannedApis.ToImmutable(), symbolDisplayFormat),
```

Comment:
Move the `ToImmutable` call outside the operation callback?

Reply:
Fixed (via an extract method)